### DISCUSSION TITLE
How does Arrow or Notes System works? (Give technical explanation)

### DISCUSSION THREAD

**Discussion by Poinet** [FunkinCrew/Funkin#2092] [Q&A]
I am a programmar, I was very interested in how the arrow or notes system works in Friday Night Funkin. I would like a technical explanation (i.e how arrow or note system was programmatically made. What they use to achieve this, like arrays, etc). As well as holding notes, how was a holding note programmatically made?

A technical explanation is appreciated, always help out other programmars, you know haha.

**Comment by Snirozu** [comment]
when a player enters a song, the game generates notes and adds them to the `unspawnNotes` array that is sorted from lowest to highest note song position
after it's generated a `update()` function is called every frame that:
1. checks if first note in `unspawnNotes` is 1500ms ahead of current song position and if it is this note is moved to the `notes` array (the `notes` array renders notes)
2. calls `keyShit()` function

the `keyShit()` function (basically input system):
1. sorts `notes` array from the lowest to highest note song position
2. gets the first note in `notes` and checks if a key was just pressed for it and if it is this note gets judged
3. iterates through sustain notes in `notes` and checks if a note can be hit (which should be equal or lower to current song position) and if a key for this note is pressed, if it is this note gets judged

i skipped some less important shit but basically this is how input system works
